#Creator: Yue Zhang

#The purpose of this script is to compare the community composition of the positive control samples (zymo community)
#across:
#1) theoretical values provided by zymo
#2) values generated by blasting reads against the 16s gene sequences of the strains included
#3) values generated using the same classification procedure as the other samples (Greengenes)

#load required packages
packages_to_load = c('stringr','phyloseq','vegan','tidyr','ggplot2','qiime2R',
                     'dplyr','ANCOMBC','usedist','rioja','ggpubr','spaa','ape',
                     'cowplot','ggforestplot','GGally','scales','metagMisc','ggsci','rstatix',
                     'wesanderson','iCAMP')
lapply(packages_to_load,require, character.only = T)

#generating color palette
col1 = wes_palette("Rushmore1")
col2 = wes_palette("GrandBudapest2")
col3 = wes_palette('Darjeeling2')
col21 <- rev(c("tomato1","darkblue","turquoise1","lightblue","darkred","mediumblue","purple","bisque","greenyellow","yellow","violetred2","darkgreen","darkgoldenrod1","deeppink3","cadetblue4","orchid2","seagreen3","purple4","dodgerblue2","red","gray27"))

#read the best blast results
p.zymo<-qza_to_phyloseq(
  features="R_analysis_input/zymo_table.qza",
  taxonomy="R_analysis_input/zymo_taxonomy_maxxacepts1.qza")

taxa.bl = as.data.frame(tax_table(p.zymo))
taxa.bl = subset(taxa.bl,select=c(Kingdom),drop = F)
names(taxa.bl)[1] = 'Genus'
#only keep the Genus for comparison
taxa.bl$Genus = sub("_.*","",taxa.bl$Genus)

#parse the classifier results
pr<-qza_to_phyloseq(
  features="R_analysis_input/rarefied_table.qza",
  tree="R_analysis_input/rooted-tree.qza",
  taxonomy="R_analysis_input/taxonomy.qza",
  metadata="R_analysis_input/metadata_rockcreek_16s_160.txt")

pr.zymo = subset_samples(pr, type == 'zymo')
pr.asv = as.data.frame(otu_table(pr.zymo))
pr.asv = subset(pr.asv, rowSums(pr.asv)>0)
taxa.cl = as.data.frame(tax_table(pr.zymo))
taxa.cl = subset(taxa.cl, rownames(taxa.cl) %in% rownames(pr.asv), select = c(Family,Genus))
#it looks like that the classifier wasn't able to classify salmonella and escherichia, instead they were all under the family of Enterobacteriaceae

taxa.cl$Genus[taxa.cl$Family == 'Enterobacteriaceae'] = 'unclassified Enterobacteriaceae'
pr.asv.taxa = merge(x = pr.asv, y = subset(taxa.cl, select = 'Genus', drop = F), 
                    by = 0)
rownames(pr.asv.taxa) = pr.asv.taxa$Row.names
pr.asv.taxa = pr.asv.taxa[,-1]
pr.asv.taxa = aggregate(.~`Genus`,pr.asv.taxa, sum)
pr.asv.taxa[,-1] = decostand(pr.asv.taxa[,-1],2,method='total')
pr.asv.taxa = pr.asv.taxa[order(rowSums(pr.asv.taxa[,-1]),decreasing = T),]
pr.asv.taxa.7 = bind_rows(pr.asv.taxa[1:7,],colSums(pr.asv.taxa[8:nrow(pr.asv.taxa),-1]))
pr.asv.taxa.7$Genus[8] = 'Others'
rownames(pr.asv.taxa.7) = pr.asv.taxa.7$Genus                          

names(pr.asv.taxa.7)[2:3] = c('group1 Greengenes','group2 Greengenes')
#read in the best blast results
bb = read.csv('R_analysis_input/zymo.csv')
bb= bb[,-1]
names(bb) = c('Genus','group1 BLAST','group2 BLAST','theoretical')
bb$Genus = sub('_.*','',bb$Genus)
all = merge(x = pr.asv.taxa.7,y = bb,
            by = 'Genus', all.x = T, all.y = T)
all[is.na(all)] = 0

all.s = gather(all,'Sample','Frequency',-Genus)
all.s$Genus = factor(all.s$Genus, levels =c('Bacillus','Enterococcus','Lactobacillus',
                                            'Listeria','Pseudomonas','Staphylococcus',
                                            'unclassified Enterobacteriaceae','Salmonella','Escherichia',
                                            'Others'))

ggplot(data = all.s, mapping = aes(x = `Sample`, y = `Frequency`,fill = Genus)) +
  geom_bar(position = 'fill', stat="identity",color = 'black')+
  scale_fill_npg()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.text.x = element_text(angle = 90,size=10, vjust = 0.1, face = 'bold', color = 'black'),
        axis.text.y = element_text(size=15, face = 'bold', color = 'black'),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = 'bold',size =15),
        axis.line = element_line(colour = 'black'),
        legend.title=element_blank(),
        legend.text = element_text(size=12))
ggsave('R_analysis_output/positive_control/zymo.png',width = 8, height = 6)
